envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -141,38 +140,41 @@ class RBACFilterFactory(
         return methodPermissions
     }
 
-    private fun resolveClientComposites(incomingEndpoint: IncomingEndpoint, roles: List<Role>): List<ClientComposite> {
-        val clientComposites = incomingEndpoint.clients.flatMap { clientOrRole ->
-            roles.find { it.name == clientOrRole }?.clients ?: setOf(clientOrRole)
+    private fun resolveClients(
+        incomingEndpoint: IncomingEndpoint,
+        roles: List<Role>
+    ): Collection<ClientMatching> {
+        val clients = incomingEndpoint.clients.flatMap { clientOrRole ->
+            roles.find { it.name == clientOrRole.name }?.clients?.map { it } ?: setOf(clientOrRole)
         }
-        return clientComposites.sorted()
+        // sorted order ensures that we do not duplicate rules
+        return clients.toSortedSet(Comparator.comparing<ClientMatching, String> { it.name })
```

Comment:
sorting should take into account the same information as we use in:
```
val policyName = clientMatchingList.joinToString(",") { it.compositeName() }
```
otherwise, duplicated policies may happen

Reply:
Lets add test which validate this.